### PR TITLE
Migrate Utf8String perf benchmarks to benchmarkdotnet

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Ended up using `IParam` due to the complex setup that must be performed before each test case. To reviewers: does this look like a correct conversion?

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Need to figure out how not to have `[ParamsSource]` across both fields, since it's causing benchmarkdotnet to try different combinations of field values, even if the method under test doesn't use that particular field. It's causing the benchmark to take 20 minutes to run.

File: tests/Benchmarks/Utf8String/CodePointEnumeration.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Code;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using Utf8StringRealType = System.Text.Utf8.Utf8String;
+
+namespace Benchmarks.Utf8String
```
Comment (ahsonkhan) [inline tests/Benchmarks/Utf8String/CodePointEnumeration.cs]: `Benchmarks.SystemTextUtf8String` (with the folder name changed to `Benchmarks/SystemTextUtf8String/`), similar to https://github.com/dotnet/corefxlab/pull/2249/files#diff-9803e8894ed7238070c254047cf81cf7R13?

File: tests/Benchmarks/Utf8String/ConstructFromString.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            public object Value { get; }
+
+            public string ToSourceCode()
+            {
+                string valueAsString = (string)Value;
+
+                StringBuilder sb = new StringBuilder("\"");
+                foreach (char ch in valueAsString)
+                {
+                    sb.AppendFormat(CultureInfo.InvariantCulture, "\\u{0:X4}", (uint)ch);
```
Comment (ahsonkhan) [inline tests/Benchmarks/Utf8String/ConstructFromString.cs]: Can you explain, in comments, why we format the string this way for `ToSourceCode()`?

File: tests/Benchmarks/Utf8String/CodePointEnumeration.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public partial class Utf8String
```
Comment (ahsonkhan) [inline tests/Benchmarks/Utf8String/CodePointEnumeration.cs]: Rename this class to CodePointEnumeration and then we can remove `Utf8StringRealType`

File: tests/Benchmarks/Utf8String/ConstructFromString.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+using BenchmarkDotNet.Code;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using Utf8StringRealType = System.Text.Utf8.Utf8String;
+
+namespace Benchmarks.Utf8String
+{
+    public partial class Utf8String
```
Comment (ahsonkhan) [inline tests/Benchmarks/Utf8String/ConstructFromString.cs]: Similarly, rename this to ConstructFromString?

**Comment by ahsonkhan** [comment]
https://ci.dot.net/job/dotnet_corefxlab/job/master/job/ubuntu16.04_debug_prtest/1826/consoleFull#-8536224876a086b3e-df04-41d2-bc4d-43e8f9406d07

FYI @JeremyKuhne, @MisinformedDNA, @krwq 

We should track and fix this intermittent failure (or maybe even disable the test for now).

From https://github.com/dotnet/corefxlab/issues/420
> The three current tests rely on Thread.Sleep and, thus, race conditions.

File: tests/Benchmarks/Utf8String/CodePointEnumeration.cs (hunk 1 above)
Comment (ahsonkhan) [reply tests/Benchmarks/Utf8String/CodePointEnumeration.cs]: I am fine with that.

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test Innerloop Ubuntu16.04 Debug Build and Test

**Comment by GrabYourPitchforks** [comment]
Waiting for checks to go green. Thanks @ahsonkhan and @adamsitnik for all the discussion! :)

**Comment by adamsitnik** [comment]
I am going to merge it now because the benchmarks are not run as part of CI so there is no need  to wait for OSX build if it can be compiled on other OSes.

**Pull request merged**